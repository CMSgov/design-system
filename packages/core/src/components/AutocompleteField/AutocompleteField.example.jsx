/* eslint-disable react/display-name */
import AutocompleteField from './AutocompleteField';
import React from 'react';

export default function() {
  return (
    <div>
      <AutocompleteField
        items={[
          { name: 'Cook County, IL' },
          { name: 'Cook County, MD' },
          { name: 'Cook County, TN' }
        ]}
        labelText="What zip code did this person live during 2017?"
        onChange={selectedItem => console.log(selectedItem)}
        onStateChange={changes => {
          if (changes.type === '__autocomplete_keydown_enter__') {
            console.log('Yep, completed the process!');
          }
        }}
      />
    </div>
  );
}
