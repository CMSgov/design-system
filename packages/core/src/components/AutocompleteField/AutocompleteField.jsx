import Button from '../Button/Button';
import Downshift from 'downshift';
import FormLabel from '../FormLabel/FormLabel';
import PropTypes from 'prop-types';
import React from 'react';
import noop from 'nooop';
import uniqueId from 'lodash.uniqueid';

/**
 * An `AutocompleteField` component renders a typeahead input field, form label, and a clear input link.
 */
export class AutocompleteField extends React.PureComponent {
  constructor(props) {
    super(props);

    this.id = uniqueId('autocomplete_');
  }

  render() {
    const {
      clearAriaLabel,
      clearInputText,
      disabled,
      fieldRef,
      items,
      itemToString,
      labelHint,
      labelText,
      onChange,
      onStateChange,
      ...fieldProps
    } = this.props;

    return (
      <Downshift
        itemToString={itemToString}
        onChange={onChange}
        onStateChange={onStateChange}
        render={({
          clearSelection,
          getInputProps,
          getItemProps,
          highlightedIndex,
          isOpen,
          inputValue
        }) => (
          <div className="ds-u-display--flex ds-u-flex-wrap--wrap ds-c-autocomplete">
            <FormLabel
              className="ds-u-margin-top--0"
              labelClassName="ds-c-autocomplete__label"
              component="label"
              fieldId={this.id}
              hint={labelHint}
            >
              {labelText}
            </FormLabel>

            <input
              {...getInputProps({
                className: 'ds-c-field ds-c-autocomplete__input',
                disabled: disabled,
                id: this.id,
                ref: fieldRef,
                ...fieldProps
              })}
            />

            {isOpen ? (
              <div className="ds-u-border--1 ds-u-padding--1 ds-c-autocomplete__list">
                <ul className="ds-c-list--bare">
                  {items
                    .filter(
                      item =>
                        !inputValue ||
                        item.name
                          .toLowerCase()
                          .includes(inputValue.toLowerCase())
                    )
                    .map((item, index) => (
                      <li
                        className={
                          highlightedIndex === index
                            ? 'ds-u-padding--1 ds-c-autocomplete__list-item ds-c-autocomplete__list-item--active'
                            : 'ds-u-padding--1 ds-c-autocomplete__list-item'
                        }
                        key={item.id}
                        {...getItemProps({ item })}
                      >
                        {itemToString(item)}
                      </li>
                    ))}
                </ul>
              </div>
            ) : null}

            <Button
              aria-label={clearAriaLabel}
              className="ds-u-margin-left--auto ds-u-padding-right--0 ds-c-autocomplete__button"
              href="javascript:void(0);"
              onClick={clearSelection}
              size="small"
              variation="transparent"
            >
              {clearInputText}
            </Button>
          </div>
        )}
      />
    );
  }
}

AutocompleteField.defaultProps = {
  clearAriaLabel: 'Clear typeahead and search again',
  clearInputText: 'Search again',
  itemToString: noop,
  labelHint:
    'This is an autocomplete field. Begin typing to search for relevant information. The number of results will be updated as you type.',
  labelText: 'Generic autocomplete label',
  onChange: noop,
  onStateChange: noop
};

AutocompleteField.propTypes = {
  /**
   * Screenreader-specific label for the Clear input link. Intended to provide a longer, more descriptive explanation of the link's behavior.
   */
  clearAriaLabel: PropTypes.string,
  /**
   * Clear link text that will appear on the page as part of the rendered component
   */
  clearInputText: PropTypes.string,
  disabled: PropTypes.bool,
  /**
   * https://github.com/paypal/downshift#itemtostring
   *
   * Used to determine the string value for the selected item (which is used to compute the `inputValue`.
   */
  itemToString: PropTypes.func.isRequired,
  /**
   * Array of objects used to populate the suggestion list that appears below the input as users type. This array of objects is intended for an async data callback, and should conform to the prescribed shape to avoid errors.
   */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.name
    })
  ).isRequired,
  /**
   * Access a reference to the `input` element
   */
  fieldRef: PropTypes.func,
  /**
   * Screenreader-specific instructions for using the autocomplete. The hint is visually hidden, but will be read out to assistive devices when the input receives keyboard focus.
   */
  labelHint: PropTypes.string,
  /**
   * Text string label for the `<FormLabel>` component
   */
  labelText: PropTypes.string.isRequired,
  /**
   * https://github.com/paypal/downshift#onchange
   *
   * Called when the user selects an item and the selected item has changed. Called with the item that was selected and the new state of `downshift`.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * https://github.com/paypal/downshift#onstatechange
   *
   * This function is called anytime the internal state changes. This can be useful if you're using downshift as a "controlled" component, where you manage some or all of the state (e.g. `isOpen, selectedItem, highlightedIndex`, etc) and then pass it as props, rather than letting downshift control all its state itself.
   */
  onStateChange: PropTypes.func.isRequired
};

export default AutocompleteField;
