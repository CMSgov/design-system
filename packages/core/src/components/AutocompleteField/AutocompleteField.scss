// Form-specific variables can be found in _variables.forms
@import '@cmsgov/design-system-support/src/settings/index';

/*
Autocomplete field

@uswds https://standards.usa.gov/components/form-controls/#text-input

Text fields allow users to enter any combination of letters, numbers, or symbols of their choosing (unless otherwise restricted). Text fields can span single or multiple lines.

Markup:
<label class="ds-c-label ds-u-margin-top--0" for="input-firstname">First name</label>
<input class="ds-c-field" id="input-firstname" name="firstname" type="text">
<label class="ds-c-label" for="input-lastname">
  <span class="ds-u-font-weight--bold">Last name</span>
  <span class="ds-c-field__hint ds-u-color--error" role="alert">
    Please enter your last name
  </span>
  <span class="ds-c-field__hint">Helpful hint text</span>
</label>
<input class="ds-c-field ds-c-field--error" id="input-lastname" name="lastname" type="text" />
<label class="ds-c-label" for="input-username">
  Username
  <span class="ds-c-field__hint">Example of a success field state</span>
</label>
<input class="ds-c-field ds-c-field--success" id="input-username" name="username" type="text" value="johnsnow">
<label class="ds-c-label" for="input-bio">Bio</label>
<textarea class="ds-c-field" id="input-bio" name="bio" rows="5"></textarea>
<label class="ds-c-label" for="input-id-num">
  ID Number
  <span class="ds-c-field__hint">
    Example of a disabled field
  </span>
</label>
<input class="ds-c-field ds-c-field--inverse" id="input-id-num" name="id-num" type="text" value="1234" disabled>

Style guide: components.autocomplete-field
*/

// input
.ds-c-autocomplete {
  max-width: 460px;

  &__label {
    + .ds-c-field__hint {
      border: 0;
      clip: rect(0, 0, 0, 0);
      height: 1px;
      overflow: hidden;
      padding: 0;
      position: absolute;
      width: 1px;
      word-wrap: normal;
    }
  }

  &__list {
    width: 100%;
  }

  &__list-item--active {
    background-color: $color-primary-alt-darkest;
    color: $color-white; 
  }
}

/*
`<Autocompletefield>`

@react-component AutocompleteField

Style guide: components.autocomplete-field.react
*/

/*
---

## When to use

- If you can’t reasonably predict a user’s answer to a prompt and there might be wide variability in users’ answers.
- When using another type of input will make answering more difficult. For example, birthdays and other known dates are easier to type in than they are to select from a calendar picker.
- When users want to paste in a response.

## When to consider alternatives

- When users are choosing from a specific set of options. Consider [checkboxes, radio buttons]({{root}}/components/choice), or a [select menu]({{root}}/components/select) in these cases.

## Guidance

- Don't use placeholder text in form fields. Use hint text instead, if you need to provide contextual information. Placeholder text disappears after a user types a value, therefore users will no longer have that text available when they need to review their entries. People who have cognitive or visual disabilities have additional problems with placeholder text.
- The length of the text field provides a hint to users as to how much text to write. Do not require users to write paragraphs of text into a single-line `input` box; use a `textarea` instead.
- Text fields are among the easiest type of input for desktop users but are more difficult for mobile users.
- Only show error messages or styling after a user has interacted with a particular field.

**[View the "Forms" guidelines for additional guidance and best practices.]({{root}}/guidelines/forms/)**

## Accessibility

- Group each set of thematically related controls in a `fieldset` element. Use the `legend` element to offer a label within each `fieldset`. The `fieldset` and `legend` elements make it easier for screen reader users to navigate the form.
- Keep your form blocks in a vertical pattern. This approach is ideal, from an accessibility standpoint, because of limited vision that makes it hard to scan from left to right.

## Theming

The following Sass variables can be overridden to theme a field:

- `$color-focus`
- `$color-focus-inverse`
- `$focus-shadow`
- `$focus-shadow-inverse`
- `$input-border-color`
- `$input-border-color-inverse`
- `$input-border-radius`
- `$input-line-height`
- `$input-border-width`
- `$input-padding`

## Related patterns

- [Date field]({{root}}/patterns/date-field/)

## Learn more

- [Form Guidelines]({{root}}/guidelines/forms/)
- ["Placeholders in Form Fields Are Harmful"](https://www.nngroup.com/articles/form-design-placeholders/)
- [Asking for a date of birth](https://designnotes.blog.gov.uk/2013/12/05/asking-for-a-date-of-birth/)
- [GOV.UK text boxes discussion](https://paper.dropbox.com/doc/Text-boxes-8NLlgz9tjR8OzTz2N9wE3)
- [Four steps for choosing form elements on the Web (PDF)](http://www.formsthatwork.com/files/Articles/dropdown.pdf)

Style guide: components.autocomplete-field.guidance
*/
