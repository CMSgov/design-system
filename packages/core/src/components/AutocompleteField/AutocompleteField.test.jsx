import { mount, shallow } from 'enzyme';
import AutocompleteField from './AutocompleteField';
import React from 'react';
import renderer from 'react-test-renderer';

function render(customProps = {}, deep = false) {
  const props = Object.assign(
    {
      items: [{ id: 'kRf6c2fY', name: 'Cook County, IL' }]
    },
    customProps
  );
  const component = <AutocompleteField {...props} />;

  return {
    props: props,
    wrapper: deep ? mount(component) : shallow(component)
  };
}

describe('AutocompleteField', () => {
  it('renders AutocompleteField component', () => {
    const data = render({}, false);
    const wrapper = data.wrapper;
    const inst = wrapper.instance();

    expect(inst).toBeInstanceOf(AutocompleteField);
  });

  it('returns correct default props', () => {
    const data = render({}, true);
    const wrapper = data.wrapper;

    expect(wrapper.prop('clearAriaLabel')).toBe(
      'Clear typeahead and search again'
    );
    expect(wrapper.prop('clearInputText')).toBe('Search again');
    expect(wrapper.prop('disabled')).toBe(undefined);
    expect(wrapper.prop('itemToString')).toBe(undefined);
    expect(wrapper.prop('fieldRef')).toBe(undefined);
    expect(wrapper.prop('labelHint')).toBe(
      'This is an autocomplete field. Begin typing to search for relevant information. The number of results will be updated as you type.'
    );
    expect(wrapper.prop('labelText')).toBe('Generic autocomplete label');
    expect(wrapper.prop('onChange')).toBe(undefined);
    expect(wrapper.prop('onStateChange')).toBe(undefined);
  });

  it('renders expected elements', () => {
    const data = render({}, true);
    const wrapper = data.wrapper;
    const downshift = wrapper.find('Downshift');

    expect(downshift.find('label').exists()).toBe(true);
    expect(downshift.find('span.ds-c-autocomplete__label').exists()).toBe(true);
    expect(downshift.find('span.ds-c-field__hint').exists()).toBe(true);
    expect(downshift.find('input').exists()).toBe(true);
    expect(downshift.find('a').exists()).toBe(true);
  });

  it('renders default class names', () => {
    const data = render({}, true);
    const wrapper = data.wrapper;
    const downshift = wrapper.find('Downshift');

    expect(downshift.find('label').hasClass('ds-c-label')).toBe(true);
    expect(downshift.find('input').hasClass('ds-c-autocomplete__input')).toBe(
      true
    );
    expect(downshift.find('a').hasClass('ds-c-autocomplete__button')).toBe(
      true
    );
  });

  it('allows default props to be overridden', () => {
    const data = render(
      {
        clearAriaLabel: 'New ARIA label'
      },
      true
    );
    const wrapper = data.wrapper;

    expect(wrapper.prop('clearAriaLabel')).toBe('New ARIA label');
  });

  it('allows custom props to be passed to the input', () => {
    const data = render(
      {
        example_prop: 'Example string'
      },
      true
    );
    const wrapper = data.wrapper;
    const input = wrapper.find('input');
    const label = wrapper.find('label');

    expect(input.prop('example_prop')).toBe('Example string');
    expect(label.prop('example_prop')).toBe(undefined);
  });

  it('renders a snapshot', () => {
    const tree = renderer
      .create(
        <AutocompleteField
          items={[{ id: 'kRf6c2fY', name: 'Cook County, IL' }]}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
