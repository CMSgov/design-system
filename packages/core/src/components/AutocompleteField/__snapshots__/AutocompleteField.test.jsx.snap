// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteField renders a snapshot 1`] = `
<div
  className="ds-u-display--flex ds-u-flex-wrap--wrap ds-c-autocomplete"
>
  <label
    className="ds-c-label ds-u-margin-top--0"
    htmlFor="autocomplete_7"
  >
    <span
      className="ds-c-autocomplete__label"
    >
      Generic autocomplete label
    </span>
    <span
      className="ds-c-field__hint"
    >
      This is an autocomplete field. Begin typing to search for relevant information. The number of results will be updated as you type.
    </span>
  </label>
  <input
    aria-activedescendant={null}
    aria-autocomplete="list"
    aria-expanded={false}
    autoComplete="off"
    className="ds-c-field ds-c-autocomplete__input"
    disabled={undefined}
    id="autocomplete_7"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    role="combobox"
    value=""
  />
  <a
    aria-label="Clear typeahead and search again"
    className="ds-c-button ds-c-button--transparent ds-c-button--small ds-u-margin-left--auto ds-u-padding-right--0 ds-c-autocomplete__button"
    href="javascript:void(0);"
    onClick={[Function]}
  >
    Search again
  </a>
</div>
`;
